cpp/src/arrow/dataset/file_parquet.h

```diff
@@ -154,23 +154,27 @@ class ARROW_DS_EXPORT ParquetFileFragment : public FileFragment {
   Result<FragmentVector> SplitByRowGroup(const std::shared_ptr<Expression>& predicate);
 
   /// \brief Return the RowGroups selected by this fragment.
-  const std::vector<int>& row_groups() const { return row_groups_; }
+  const std::vector<int>& row_groups() const {
+    if (row_groups_) return *row_groups_;
+    static std::vector<int> empty;
+    return empty;
```

Comment:
Shouldn't this be a vector of <0, 1, 2, ..num_row_groups> ? Because if `row_groups_` is not set, it's meaning all row groups?

Reply:
No; in c++ this property may be spuriously empty if you haven't called `EnsureCompleteMetadata()` yet.

Reply:
Exactly